pyanaconda/modules/subscription/runtime.py

```diff
@@ -580,3 +602,327 @@ def run(self):
+
+    @property
+    def name(self):
+        return "Restore RHSM configuration after Satellite provisioning"
+
+    def run(self):
+        "Restore the full RHSM configuration back to clean values."""
+        # the SetAll() RHSM DBus API requires a dict of variants
+        config_dict = {}
+        for key, value in self._rhsm_configuration.items():
+            # if value is present in request, use it
+            config_dict[key] = get_variant(Str, value)
+        self._rhsm_config_proxy.SetAll(config_dict, "")
+
+
+class RegisterAndSubscribeTask(Task):
+    """Register and subscribe the installation environment.
+
+    NOTE: A separate installation task make sure all the subscription related tokens
+          and configuration files are transferred to the target system, to keep
+          the machine subscribed also after installation.
+
+          In case of registration to a Satellite instance another installation task
+          makes sure the system stays registered to Satellite after installation.
+    """
+
+    def __init__(self, rhsm_observer, subscription_request, system_purpose_data,
+                 registered_callback, registered_to_satellite_callback,
+                 subscription_attached_callback, subscription_data_callback,
+                 config_backup_callback):
+        """Create a register-and-subscribe task.
+
+        :param rhsm_observer: DBus service observer for talking to RHSM
+        :param subscription_request: subscription request DBus struct
+        :param system_purpose_data: system purpose DBus struct
+
+        :param registered_callback: called when registration tasks finishes successfully
+        :param registered_to_satellite_callback: called after successful Satellite provisioning
+        :param subscription_attached_callback: called after subscription is attached
+        :param subscription_data_callback: called after subscription data is parsed
+        :param config_backup_callback: called when RHSM config data is ready to be backed up
+
+        :raises: SatelliteProvisioningError if Satellite provisioning fails
+        :raises: RegistrationError if registration fails
+        :raises: SubscriptionError if subscription fails to attach
+        """
+        super().__init__()
+        self._rhsm_observer = rhsm_observer
+        self._subscription_request = subscription_request
+        self._system_purpose_data = system_purpose_data
+        # callback for nested tasks
+        self._registered_callback = registered_callback
+        self._registered_to_satellite_callback = registered_to_satellite_callback
+        self._subscription_attached_callback = subscription_attached_callback
+        self._subscription_data_callback = subscription_data_callback
+        self._config_backup_callback = config_backup_callback
+
+    @property
+    def name(self):
+        return "Register and subscribe"
+
+    def _provision_system_for_satellite(self):
```

Comment:
It would be great to split this to even smaller methods. Right now, if I'm reading correctly, here are three unrelated things which shouldn't be so hard to split.
I see that you are only creating proxies and running tasks but IMHO it would be more readable having these abstracted.

Reply:
I think splitting the proxy stuff to a separate task makes sense. Splitting the rest seems to not add much benefit for a lot of boilerplate as indeed, it's just running tasks in the correct order.